### PR TITLE
Migrate to April 12, 2016 snapshot

### PR DESCRIPTION
**Pull request by paulofaria**
The biggest API change in this new snapshot is that `_` is not implicit for first parameters, so we have to add them explicitly to maintain the former behaviour.

### CODE REVIEW

File: Sources/Stream.swift [hunk 1]
```diff
@@ -22,7 +22,7 @@ extension Sending {
 }
 
 extension Sending {
-    public func send(data: Data) throws {
+    public func send(_ data: Data) throws {
```
Comment (Joannis) [inline Sources/Stream.swift:25]: Should be separate from this pull request I think... or is it required now?

Comment (paulofaria) [inline Sources/Stream.swift:25]: This is required now.

File: Sources/Data.swift [hunk 2]
```diff
@@ -28,7 +28,7 @@ extension Data: RangeReplaceableCollection, MutableCollection {
     }
 
     #if swift(>=3.0)
-        public mutating func replaceSubrange<C : Collection where C.Iterator.Element == Byte>(subRange: Range<Int>, with newElements: C) {
+        public mutating func replaceSubrange<C : Collection where C.Iterator.Element == Byte>(_ subRange: Range<Int>, with newElements: C) {
```
Comment (Obbut) [inline Sources/Data.swift:31]: I think this would fit Swift and the standard library better with the current snapshot as `replace(subRange: ..., with: )`, but that may be better suited for a different PR

Edit: This line actually comes from the `RangeReplaceableCollection` protocol I see now, but still this would be a thing to look at for the rest of the API.

Comment (tanner0101) [inline Sources/Data.swift:31]: @Obbut upgrading Swift versions is going to force us to do a minor bump anyway, so we might as well get this out of the way.

File: Sources/AsyncConnection.swift [hunk 3]
```diff
@@ -3,7 +3,7 @@ public protocol AsyncConnection: AsyncStream {
 }
 
 extension AsyncConnection {
-    public func open(result: (Void throws -> AsyncConnection) -> Void) throws {
+    public func open(_ result: (Void throws -> AsyncConnection) -> Void) throws {
```
Comment (dreymonde) [inline Sources/AsyncConnection.swift]: I don’t like how `result` looks. Can be more clear. Maybe we should rename it to `completion` (which is, I guess, more familiar for the vast majority of Cocoa developers) and don’t substitute with `_`. So it will be like:

``` swift
connection.open(completion: { … })
```

or

``` swift
connection.open {
    ...
}
```

Comment (tanner0101) [inline Sources/AsyncConnection.swift]: Yeah it should probably be `completion`, `closure`, or `callback` something... We should be consistent. The `result` closure seems to be the input to the `completion` closure.

Comment (paulofaria) [inline Sources/AsyncConnection.swift]: > The result closure seems to be the input to the completion closure.

makes sense! I think `completion` is good. `closure` or `callback` are type information not `role` information. `completion` does a good job explaining the intent.

Comment (Danappelxx) [inline Sources/AsyncConnection.swift]: I'm assuming once we have generic type aliases we'll just have something like
`typealias Completion<T> = (Void throws -> T) -> Void`

Also, I'm not a huge fan of that closure anyway - it's really hard to parse without seeing a good example or thinking about it for a while first. Perhaps we should make a wrapper around it (and then we can take advantage of generics)? LMK if this is something that I should raise discussion about separately.

Comment (paulofaria) [inline Sources/AsyncConnection.swift]: I think generic typealias is already a thing. Have to test. But.. would generics be a good idea? Think there could be disadvantages.

Comment (Danappelxx) [inline Sources/AsyncConnection.swift]: I think it's much clearer what's going on...

``` swift
func open(_ result: (Void throws -> AsyncConnection) -> Void) throws
```

vs

``` swift
func open(_ completion: Completion<AsyncConnection>) throws
```

Comment (dreymonde) [inline Sources/AsyncConnection.swift]: Yes, I think we shouldn’t have typealiases in the library of standards. Especially these kinds of typealiases.

Comment (tanner0101) [inline Sources/AsyncConnection.swift]: > I think we shouldn’t have typealiases in the library of standards

That's fair. They make it much more difficult to reason about what is happening at first glance because you don't know what the typealias equals.

Also `completion` is not a perfect name. What we have is a closure that is handed a result closure. This seems like a more specific type of completion callback. A typealias should make it clear there are 2 closures involved.

Comment (dreymonde) [inline Sources/AsyncConnection.swift]: As @paulofaria mentioned, argument label must describe role, not the type. `callback` is reasonable because it will be called after the completion of needed action, no matter if that was successful. For example, `NSURLSession` takes `completionHandler`, which gets called in every possible case - even if the error occurs, the same handler deals with it.
So, what I'm saying is that `completion` describes the role of the argument, and from its type it's clear that it's a closure that wraps up another closure. And `typealias`, I am sure, will only make it much less readable.

Comment (tanner0101) [inline Sources/AsyncConnection.swift]: @dreymonde I agree. I wasn't arguing against the label, just the name of the typealias if we were to use one. I think the label is fine. :)

Comment (dreymonde) [inline Sources/AsyncConnection.swift]: Oops, sorry then :laughing: 
I think we should avoid the use of `typealias` in the vast majority of cases.

Comment (Danappelxx) [inline Sources/AsyncConnection.swift]: @dreymonde what did you mean by this?

> especially these kinds of tupealiases

Personally I think that tupealiases are only ok in two places: where they seriously reduce boilerplate, or when they make things more clear. I think in this case it satisfies both constraints and it makes a lot of sense to use it.

Comment (dreymonde) [inline Sources/AsyncConnection.swift]: @Danappelxx I’m okay with typealiases, but not in the library of standards. `Completion<AsyncConnection>` may be clear once you get it, but at first glance you will feel like «wait, what’s the `Completion`? Is that a `struct` or something?» And that doesn’t feels right for the library that intended to be used really really often.

Mainly, my only argument is that «Swift standard library wouldn’t have anything like that». I don’t know if that’s enough, but that’s my point of view 😄

Comment (Danappelxx) [inline Sources/AsyncConnection.swift]: That's fair. My only argument is that if we're using the same exact pattern _everywhere_, it makes sense (to me) to create a typealias.
